### PR TITLE
Enable `linear` for nvFuser

### PR DESCRIPTION
**Pull request by Priya2698**
## What does this PR do?

Enables linear in nvFuser. Part of resolving https://github.com/NVIDIA/Fuser/issues/2053

### CODE REVIEW

File: thunder/tests/test_nvfuser.py [hunk 1]
```diff
@@ -850,3 +850,35 @@ def get_num_fusions(cfn):
...
+
+        out = compiled_func(a, b, bias)
+        traces = thunder.last_traces(compiled_func)
+        fusions = examine.get_fusions(traces[-1])
+        nv_version = nvfuser_version()
+
+        expected_fusions = 1 if nv_version >= "0.2.3" else 0
+
+        assert len(fusions) == expected_fusions
+        assert torch.allclose(out, torch.nn.functional.linear(a, b, bias))
```
Comment (crcrpar) [inline thunder/tests/test_nvfuser.py:884]: nitpicking
```suggestion
        torch.testing.assert_close(out, torch.nn.functional.linear(a, b, bias))
```

File: thunder/executors/nvfuserex_impl.py [hunk 2]
```diff
@@ -2194,3 +2194,41 @@ def map_inside_replacement(x: Any) -> None:
...
+    if nv_version < LooseVersion("0.2.3"):
+        return False
+
+    enable_linear: None | bool = get_compile_option("nv_enable_linear", "Enable nvFuser matmul.")
+    if not enable_linear:
+        return False
+    # Verify linear inputs and bias (optional) are supported tensors.
+    if not are_supported_tensors(a, b):
+        return False
+    if bias is not None and not is_supported_tensor(bias):
```
Comment (jjsjann123) [inline thunder/executors/nvfuserex_impl.py:2209]: nitpick: We can probably return false in `is_supported_tensor(x)` when `x` is not a tensorproxy.

File: thunder/executors/nvfuserex_impl.py [hunk 3]
```diff
@@ -2194,3 +2194,41 @@ def map_inside_replacement(x: Any) -> None:
...
+        return False
+
+    # nvFuser only supports 2D inputs in v0.2.3.
+    if not (a.ndim == b.ndim and a.ndim == 2):
```
Comment (jjsjann123) [inline thunder/executors/nvfuserex_impl.py:2213]: prim b.ndim is always 2. `thunder/core/prims.py`. We can drop the first check.

File: thunder/executors/nvfuserex_impl.py [hunk 4]
```diff
@@ -2194,3 +2194,41 @@ def map_inside_replacement(x: Any) -> None:
...
+    bias: TensorProxy | None,
+    *,
+    fd: FusionDefinition,
+    lc_to_nv_map: dict,
+) -> Any:
+    nva = getnv(a, fd, lc_to_nv_map)
+    nvb = getnv(b, fd, lc_to_nv_map)
+    nvbias = None
+    if bias is not None:
+        nvbias = getnv(bias, fd, lc_to_nv_map)
```
Comment (jjsjann123) [inline thunder/executors/nvfuserex_impl.py:2230]: nitpick
```suggestion
    nvbias = None if bias is None else getnv(bias, fd, lc_to_nv_map)
```

File: thunder/tests/test_nvfuser.py [hunk 5]
```diff
@@ -850,3 +850,35 @@ def get_num_fusions(cfn):
...
+            bias = torch.randn(n, dtype=torch_dtype, device=device)
+
+        compiled_func = thunder.jit(fn, executors_list=executor.executors_list(), nv_enable_linear=True)
+
+        out = compiled_func(a, b, bias)
+        traces = thunder.last_traces(compiled_func)
+        fusions = examine.get_fusions(traces[-1])
+        nv_version = nvfuser_version()
+
+        expected_fusions = 1 if nv_version >= "0.2.3" else 0
```
Comment (jjsjann123) [inline thunder/tests/test_nvfuser.py:881]: nitpick: wrap the str as `LooseVersion` maybe?

**Review by jjsjann123 — APPROVED** [review]
minor comments LGTM

File: thunder/executors/nvfuserex_impl.py (hunk 2 above)
Comment (Priya2698) [reply thunder/executors/nvfuserex_impl.py:2209]: I am not clear on this suggestion. We would still need to test for supported tensor only when bias is not None (which is a valid case).

**Review by t-vi — APPROVED** [review]
Thank you @Priya2698 @crcrpar @jjsjann123.

File: thunder/executors/nvfuserex_impl.py (hunk 2 above)
Comment (jjsjann123) [reply thunder/executors/nvfuserex_impl.py:2209]: Ha, sorry I wasn't thinking straight. :+1:

File: thunder/executors/nvfuserex_impl.py [hunk 6]
```diff
@@ -2194,3 +2194,39 @@ def map_inside_replacement(x: Any) -> None:
...
+        return False
+    # Verify linear inputs and bias (optional) are supported tensors.
+    if not are_supported_tensors(a, b):
+        return False
+    if bias is not None and not is_supported_tensor(bias):
+        return False
+
+    # nvFuser only supports 2D inputs in v0.2.3.
+    if not a.ndim == 2:
+        return False
```
Comment (IvanYashchuk) [inline thunder/executors/nvfuserex_impl.py:2214]: In the real GPT network, the `a` input is always batched. What are the complications of the batched case? Is there a tracking issue for this?
cc @jjsjann123, @Priya2698

File: thunder/tests/test_nvfuser.py [hunk 7]
```diff
@@ -850,3 +851,35 @@ def get_num_fusions(cfn):
     assert get_num_fusions(cfn_without_fusion) == 0
 
     nvfuserex.set_fuel(thunder.extend.FUEL_LEVEL.UNLIMITED)
+
+
+@instantiate(
+    dtypes=(thunder.float16, thunder.bfloat16), devicetypes=(devices.DeviceType.CUDA,), executors=(nvFuserExecutor,)
+)
```
Comment (IvanYashchuk) [inline thunder/tests/test_nvfuser.py:858]: Is fp32, fp64 dtype supported?

File: thunder/executors/nvfuserex_impl.py [hunk 8]
```diff
@@ -2194,3 +2194,39 @@ def map_inside_replacement(x: Any) -> None:
...
+    b: TensorProxy,
+    bias: TensorProxy | None,
+    *,
+    fd: FusionDefinition,
+    lc_to_nv_map: dict,
+) -> Any:
+    nva = getnv(a, fd, lc_to_nv_map)
+    nvb = getnv(b, fd, lc_to_nv_map)
+    nvbias = None if bias is None else getnv(bias, fd, lc_to_nv_map)
+    return fd.ops.linear(nva, nvb, nvbias)
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply thunder/executors/nvfuserex_impl.py:2229]: `linear` is a primitive in nvFuser. See https://github.com/NVIDIA/Fuser/blob/e86bab86cdcd503e72cc9e67f9b3f49d5926cc39/csrc/ops/composite.cpp#L81

File: thunder/executors/nvfuserex_impl.py (hunk 6 above)
Comment (Priya2698) [reply thunder/executors/nvfuserex_impl.py:2214]: Yes, right now, the support is limited to 2D. The relevant issues are: 
1. https://github.com/NVIDIA/Fuser/issues/2132
2. https://github.com/NVIDIA/Fuser/issues/2092
3. https://github.com/NVIDIA/Fuser/issues/2149

I am working on adding new IR nodes for `matmul`/`linear` which will remove this restriction.

File: thunder/tests/test_nvfuser.py [hunk 9]
```diff
@@ -850,3 +851,35 @@ def get_num_fusions(cfn):
...
+def test_linear(executor, device: str, dtype: dtypes.dtype):
+
+    def fn(a, b, bias=None):
+        return torch.nn.functional.linear(a, b, bias)
+
+    m, n, k = 128, 64, 32
+    torch_dtype = ltorch.to_torch_dtype(dtype)
+    a = torch.randn((m, k), dtype=torch_dtype, device=device)
+    b = torch.randn((n, k), dtype=torch_dtype, device=device)
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply thunder/tests/test_nvfuser.py:867]: I didn't use it since we only have partial support for linear at the moment, in terms of shape. Will update the tests to use the input generators once the new IR nodes go in.

File: thunder/tests/test_nvfuser.py (hunk 7 above)
Comment (Priya2698) [reply thunder/tests/test_nvfuser.py:858]: At the moment, we do have that restriction since we use fusedMultiplySum and MmaOp that only accept bf16/fp16. This restriction will also be lifted with the new IR nodes.

File: thunder/executors/nvfuserex_impl.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Priya2698) [reply thunder/executors/nvfuserex_impl.py:2229]: This will change shortly with the new IR nodes. There will be a new `linear` primitive at the C++ interface as well creating a `LinearOp`

File: thunder/executors/nvfuserex_impl.py (hunk 6 above)
Comment (Priya2698) [reply thunder/executors/nvfuserex_impl.py:2214]: Draft PR for matmul IR node: https://github.com/NVIDIA/Fuser/pull/2175
